src/datacollector/DataCollectionAttachmentManager.cs

```diff
@@ -0,0 +1,458 @@
+            {
+                attachmentRequest.WaitForCopyComplete();
+                if (attachmentRequest.Error != null)
+                {
+                    // If there was error in processing the request, lets log it.
+                    var testCaseId = dataCollectionContext.HasTestCase
+                                         ? dataCollectionContext.TestExecId.Id
+                                         : Guid.Empty;
+                    this.LogError(
+                        attachmentRequest.Error.Message,
+                        attachmentRequest.FileTransferInfo.Uri,
+                        attachmentRequest.FileTransferInfo.FriendlyName,
+                        testCaseId);
+                }
+                else
+                {
+                    // Create collectorDataEntry for each collected log.
+                    var entry =
+                        entries.FirstOrDefault(
+                            e => Uri.Equals(e.Uri, attachmentRequest.FileTransferInfo.Uri));
+                    if (entry == null)
+                    {
+                        entry = new AttachmentSet(
+                            attachmentRequest.FileTransferInfo.Uri,
+                            attachmentRequest.FileTransferInfo.FriendlyName);
+                        entries.Add(entry);
+                    }
+
+                    entry.Attachments.Add(
+                        new UriDataAttachment(
+                            new Uri(attachmentRequest.LocalFilePath),
+                            attachmentRequest.FileTransferInfo.Description));
+                }
+
+                attachmentRequest.Dispose();
+            }
+
+            return entries;
+        }
+
+        /// <inheritdoc/>
+        public void AddAttachment(FileTransferInformationExtension fileTransferInfo)
+        {
+            ValidateArg.NotNull(fileTransferInfo, nameof(fileTransferInfo));
+
+            if (string.IsNullOrEmpty(this.SessionOutputDirectory))
+            {
+                if (EqtTrace.IsErrorEnabled)
+                {
+                    EqtTrace.Error(
+                        "DataCollectionAttachmentManager.AddAttachment: Initialize not invoked.");
+                }
+
+                return;
+            }
+
+            if (fileTransferInfo != null)
+            {
+                this.AddNewFileTransfer(fileTransferInfo);
+            }
+            else
+            {
+                if (EqtTrace.IsErrorEnabled)
+                {
+                    EqtTrace.Error("DataCollectionAttachmentManager.AddAttachment: Got unexpected message of type FileTransferInformationExtension.");
+                }
+            }
+        }
+
+        /// <summary>
+        /// Dispose event object
+        /// </summary>
+        public void Dispose()
+        {
+            // Dispose all attachment requests..
+            this.AttachmentRequests.ForEach(request => request.Dispose());
```

Comment:
>AttachmentRequests [](start = 17, length = 18)

If AttachmentRequests are not required (in case of Task), we may not need Dispose.